File: backend/arkham-api/library/Entity/Arkham/Achievement.hs
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE UndecidableInstances #-}

module Entity.Arkham.Achievement (
  module Entity.Arkham.Achievement,
) where

import Arkham.Achievement.Types (Achievement)
import Data.Time.Clock
import Data.UUID (UUID)
import Database.Persist.TH
import Entity
import Entity.Arkham.Game
import Entity.User
import Json
import Orphans ()
import Relude

-- One row per user per achievement, forever. `arkhamGameId` links back to the
-- earning game and degrades to NULL when that game is deleted. `progress`
-- carries counter/checklist state for cross-playthrough achievements while
-- `earnedAt` is still null.
mkEntity
  $(discoverEntities)
  [persistLowerCase|
ArkhamAchievement sql=arkham_achievements
  Id UUID default=uuid_generate_v4()
  userId UserId OnDeleteCascade
  achievement Achievement
  earnedAt UTCTime Maybe
  arkhamGameId ArkhamGameId Maybe OnDeleteSetNull
  progress Value
  UniqueUserAchievement userId achievement
  deriving Generic Show
|]

instance ToJSON ArkhamAchievement where
  toJSON = genericToJSON $ aesonOptions $ Just "arkhamAchievement"
  toEncoding = genericToEncoding $ aesonOptions $ Just "arkhamAchievement"

File: frontend/src/arkham/types/Achievement.ts
import * as JsonDecoder from 'ts.data.json'
import { withDefault } from '@/arkham/parser'

// One row per user per achievement (entity format: {id, ...fields} with the
// "arkhamAchievement" prefix stripped). `earnedAt` is null while the row only
// tracks progress; `arkhamGameId` degrades to null when the game is deleted.
export type Achievement = {
  id: string
  userId: number | null
  achievement: string
  earnedAt: string | null
  arkhamGameId: string | null
  progress: unknown
}

export const achievementDecoder = JsonDecoder.object<Achievement>(
  {
    id: JsonDecoder.string(),
    userId: withDefault<number | null>(null, JsonDecoder.number()),
    achievement: JsonDecoder.string(),
    earnedAt: withDefault<string | null>(null, JsonDecoder.string()),
    arkhamGameId: withDefault<string | null>(null, JsonDecoder.string()),
    progress: JsonDecoder.succeed(),
  },
  'Achievement',
)
